### PR TITLE
Add SwiftUI enhancements

### PR DESCRIPTION
**Pull request by jsflax**
This PR adds the `@StateRealmObject`, `@ObservedObject`, and `@ObservedResults` property wrappers, which will allow users to more easily wield thread confined live types: `Objects`, `EmbeddedObjects`, `Lists`, or `Results`:

```swift
struct PersonView: View {
    @ObservedResults(Person.self) var results

    var body: some View {
        return NavigationView {
            List {
                ForEach(results) { person in
                    NavigationLink(destination: PersonDetailView(person: person)) {
                        Text(person.name)
                    }
                }
                .onDelete(perform: $results.remove)
            }
            .navigationBarTitle("People", displayMode: .large)
            .navigationBarItems(trailing: Button("Add") {
                $results.append(Person())
            })
        }
    }
}
```

The wrapper allows properties to be wielded normally when doing most operations, and adds bonus convenience methods for under-the-hood write transactions for each type (and their properties).

### CODE REVIEW

File: RealmSwift/SwiftUI.swift [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+                guard let (_, resolved) = resolve(wrappedValue) else {
+                    return wrappedValue[keyPath: member]
+                }
+                return resolved[keyPath: member]
+            },
+            set: { newValue in
+                guard let (realm, resolved) = resolve(wrappedValue) else {
+                    return
+                }
+                try! realm.write {
```
[the remark this answers is not in the document]
Comment (jsflax) [reply RealmSwift/SwiftUI.swift:67]: The setter cannot throw because it is used to set the wrappedValue with the nonmutating setter. What's the alternative?

File: SwiftUITests/SwiftUITests.swift [hunk 2]
```diff
@@ -0,0 +1,114 @@
+//
```
Comment (leemaguire) [inline SwiftUITests/SwiftUITests.swift:1]: needs a licence header

File: examples/ios/swift/ListSwiftUI/Views/ContentView.swift [hunk 3]
```diff
@@ -1,126 +1,148 @@
...
-                }
-            }.listStyle(GroupedListStyle())
-                .navigationBarTitle("recipes", displayMode: .large)
-                .navigationBarBackButtonHidden(true)
-                .navigationBarHidden(false)
-                .navigationBarItems(trailing: EditButton())
+        List {
+            // Using the `$` will bind the Dog List to the view
+            ForEach(dogs) { dog in
+                TextField("dog name", text: bind(dog, \.name))
```
Comment (leemaguire) [inline examples/ios/swift/ListSwiftUI/Views/ContentView.swift:68]: will this `bind` API be hard to discover? Do we also want to extend `View` to do the also?

Comment (jsflax) [reply examples/ios/swift/ListSwiftUI/Views/ContentView.swift:68]: My other thought was to put `bind` on `ThreadConfined` types, but I'm not sure if there is value in that outside of SwiftUI (it's API bloat otherwise, but maybe that's fine?). I'm open to ideas anyways.

Comment (leemaguire) [reply examples/ios/swift/ListSwiftUI/Views/ContentView.swift:68]: I'm just thinking we should extend our own domain and not another party's. But it's tricky because its SwiftUI specific. I guess I'm fine with extending `View` as its a very niche case

Comment (jsflax) [reply examples/ios/swift/ListSwiftUI/Views/ContentView.swift:68]: I actually don't think this is terribly niche, which is why I'm trying to be careful– binding to a SwiftUI view will likely be fairly common (TextFields, sliders, etc.).

Thoughts @tgoyne ?

**Comment by jsflax** [comment]
@tgoyne I've diverged from Eric's branch and will changeover to master once Eric's changes are merged, but could use another set of eyes on the actual swiftUI related changes.

File: Configuration/SwiftUITestHost.xcconfig [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+AD_HOC_CODE_SIGNING_ALLOWED = YES;
+CODE_SIGNING_REQUIRED = NO;
+
+PRODUCT_BUNDLE_IDENTIFIER = io.realm.SwiftUITestHost;
+
+OTHER_LDFLAGS[sdk=macosx*] = -framework $(REALM_UI_FRAMEWORK_$(RESOURCES_UI_FRAMEWORK_FAMILY));
+OTHER_LDFLAGS[sdk=macosx*] = -framework SwiftUI;
+LD_RUNPATH_SEARCH_PATHS[sdk=macosx*] = $(inherited) @executable_path/../Frameworks @loader_path/../Frameworks;
+
+MACOS_DEPLOYMENT_TARGET = 11.0;
```
Comment (tgoyne) [inline Configuration/SwiftUITestHost.xcconfig:13]: This needs to target 10.15 to be able to run on CI.

File: Realm.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -2850,9 +2853,15 @@
...
 				CLANG_WARN__DUPLICATE_METHOD_MATCH = YES;
+				CODE_SIGN_ENTITLEMENTS = SwiftUITestHost/SwiftUITestHost.entitlements;
+				CODE_SIGN_IDENTITY = "Apple Development";
 				CODE_SIGN_STYLE = Automatic;
+				COMBINE_HIDPI_IMAGES = YES;
 				COPY_PHASE_STRIP = NO;
 				DEBUG_INFORMATION_FORMAT = dwarf;
+				DEVELOPMENT_ASSET_PATHS = "";
+				DEVELOPMENT_TEAM = "";
+				ENABLE_HARDENED_RUNTIME = YES;
```
Comment (tgoyne) [inline Realm.xcodeproj/project.pbxproj:2864]: This is why you have to do awkward things with code signing and it should just be off unless it's actually required.

File: RealmSwift/List.swift [hunk 6]
```diff
@@ -615,23 +615,33 @@ extension List: MutableCollection {
...
-        for offset in offsets {
-            tmp.append(self[offset])
-        }
-        insert(contentsOf: tmp, at: destination)
-        for offset in offsets.reversed() {
-            var o = offset
-            if o >= destination {
-                o += tmp.count
+        if Element.self is EmbeddedObject.Type {
+            for offset in offsets {
```
Comment (tgoyne) [inline RealmSwift/List.swift:621]: Is there some reason not to use this code path for non-embedded objects as well?

File: RealmSwift/SwiftUI.swift [hunk 7]
```diff
@@ -21,13 +21,31 @@ import SwiftUI
...
+private func createBinding<T: ThreadConfined, V>(_ getter: @escaping () -> T,
+                                                 forKeyPath keyPath: ReferenceWritableKeyPath<T, V>) -> Binding<V> {
+    let lastValue = getter()[keyPath: keyPath]
+    return Binding(get: {
+        var parent = getter()
+        guard !parent.isInvalidated else {
+            return lastValue
+        }
+        if parent.isFrozen {
+            parent = try! Realm(configuration: parent.realm!.configuration).thaw(parent)!
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:34]: Can't this just be `parent.realm!.thaw(parent)`?

The force-unwrap is unsafe and this'll crash if the object was deleted after the version which the parent was frozen at.

File: RealmSwift/SwiftUI.swift [hunk 8]
```diff
@@ -142,130 +131,103 @@ extension RealmBinding: RandomAccessCollection, BidirectionalCollection, Collect
...
     public var indices: Indices {
-        return wrappedValue.indices
+        return get().indices
     }
     /// :nodoc:
     public subscript(position: Index) -> Element {
         get {
-            self.wrappedValue[position]
+            if self.get().isFrozen { throwRealmException("Cannot get values from a frozen realm") }
+            return self.get()[position]
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:156]: get() can do some nontrivial work so this should avoid calling it twice in one function.

File: RealmSwift/SwiftUI.swift [hunk 9]
```diff
@@ -300,50 +262,96 @@ private final class Box<T: RealmSubscribable & ThreadConfined>: ObservableObject
...
      */
     public init<U: Object>(_ type: U.Type, filter: NSPredicate? = nil, realm: Realm? = nil) where T == Results<U> {
-        if let realm = realm {
-            _box = StateObject(wrappedValue: Box(realm.objects(U.self).freeze()))
-        } else {
-            let realm = Environment(\.realm).wrappedValue
-            let results = filter == nil ? realm.objects(U.self) : realm.objects(U.self).filter(filter!)
-            _box = StateObject(wrappedValue: Box(results.freeze()))
-        }
+        let actualRealm = realm == nil ? try! Realm(configuration: Realm.Configuration.defaultConfiguration) : realm!
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:334]:
```suggestion
        let actualRealm = realm ?? try! Realm()
```

File: RealmSwift/List.swift (hunk 6 above)
Comment (jsflax) [reply RealmSwift/List.swift:621]: It seemed like the way for non-embedded objects was "more correct," and that this was a workaround, but I can check if that actually matters

File: RealmSwift/SwiftUI.swift [hunk 10]
```diff
@@ -0,0 +1,700 @@
...
+    /// The object to observe.
+    @ObservedObject private var storage: ObservableStorage<ObjectType>
+    /// A default value to avoid invalidated access
+    private let defaultValue: ObjectType
+
+    /// :nodoc:
+    public var wrappedValue: ObjectType {
+        get {
+            if storage.value.realm == nil {
+                // if unmangaged, return the unmanaged value
```
Comment (iandundas) [inline RealmSwift/SwiftUI.swift:594]: small typo:
```suggestion
                // if unmanaged, return the unmanaged value
```

**Comment by jsflax** [comment]
Some changes since last review:
- KVO has been added for unmanaged objects
- Optionals are now supported for `@StateRealmObject`
- `@StateRealmObject` is now exclusively for `List`, `Objects`/`EmbeddedObjects`, and `Optional` variants of the previously mentioned types. It is meant to be used in the same way as SwiftUI's `@StateObject`.
- `@ObservedRealmObject` is now exclusively for `List`, `Objects`/`EmbeddedObjects`. It is meant to be used in the same way as SwiftUI's `@ObservedObject`.
- `@FetchRealmResults` has been added to better support binding Results types to a view while allowing dynamic querying and multiple Realm support. The default realm configuration for a `@FetchRealmResults` is passed in through the environment of a view hierarchy, e.g.,:
```swift
struct View2: View {
    // the realm used here will be the realm opened from
    // `myCustomConfiguration`
    @FetchRealmResults(MyObject.self) var results

    ...
}

struct View1: View {
    var body: some View {
        View2().environment(\.realmConfiguration, myCustomConfiguration)
    }
}
```
- The example in the examples directory has been changed to the Reminders application

TODO:
- Need to update the README.
- Add one example showing multiple realm usage

@tgoyne can you PTAL?

File: RealmSwift/SwiftUI.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsflax) [reply RealmSwift/SwiftUI.swift]: Forgot to, but I still need to do this even though this comment is outdated.

File: RealmSwift/List.swift (hunk 6 above)
Comment (jsflax) [reply RealmSwift/List.swift]: I've added a test for this and collapsed the conditional.

**Comment by jsflax** [comment]
@tgoyne I've resolved all previous comments, just wrestling with the CI now.

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,6 +1,19 @@
 x.y.z Release notes (yyyy-MM-dd)
 =============================================================
 ### Enhancements
+* Add support for "thawing" objects. `Realm`, `Results`, `List` and `Object` now
+have `thaw()` methods which return a live copy of the frozen object. This enables app behvaior where a frozen object can be made live again in order to mutate values. For example, first freezing an object passed into UI view, then thawing the object in the view to update values.
+* Add `@StateRealmObject` for SwiftUI support. This property wrapper type instantiates an observable object on a View. Use in place of `SwiftUI.StateObject` for Realm `Object`, `List`, and `EmbeddedObject` types.
+* Add `@ObservedObject` for SwiftUI support. This property wrapper type subscribes to an observable object and invalidates a view whenever the observable object changes. Use in place of `SwiftUI.ObservedObject` for Realm `Object`, `List`, or `EmbeddedObject` types.
```
Comment (leemaguire) [inline CHANGELOG.md:7]: should this be `@ObservedRealmObject`?

Comment (jsflax) [reply CHANGELOG.md:7]: Yes

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,11 +1,18 @@
...
+* Add `EnvironmentValues.realm` and `EnvironmentValues.realmConfiguration` for `Realm` and `Realm.Configuration` types respectively. Values can be injected into views using the `View.environment` method, e.g., `MyView().environment(\.realmConfiguration, Realm.Configuration(fileURL: URL(fileURLWithPath: "myRealmPath.realm")))`. The value can then be declared on the example `MyView` as `@Environment(\.realm) var realm`.
+* Add `SwiftUI.Binding` extensions where `Value` is of type `Object`, `List`, or `EmbeddedObject`. These extensions expose methods for wrapped write transactions, to avoid boilerplate within views, e.g., `TextField("name", $personObject.name)` or `$personList.append(Person())`.
+* Add `Object.bind` and `EmbeddedObject.bind` for SwiftUI support. This allows you to create bindings of realm properties when a propertyWrapper is not available for you to do so, e.g., `TextField("name", personObject.bind(\.name))`.
+
+### Fixed
+* <How to hit and notice issue? what was the impact?> ([#????](https://github.com/realm/realm-cocoa/issues/????), since v?.?.?)
```
Comment (tgoyne) [inline CHANGELOG.md:14]: Incorrect merge.

File: RealmSwift/SwiftUI.swift [hunk 13]
```diff
@@ -0,0 +1,619 @@
...
+///
+/// The results use the realm configuration provided by
+/// the environment value `EnvironmentValues/realmConfiguration`.
+@available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
+@propertyWrapper public struct FetchRealmResults<ResultType>: DynamicProperty, BoundCollection where ResultType: Object & ObjectKeyIdentifiable {
+    private class Storage: ObservableStorage<Results<ResultType>> {
+        var sortDescriptor: SortDescriptor? {
+            willSet {
+                if let sortDescriptor = newValue {
+                    value = baseValue.sorted(byKeyPath: sortDescriptor.keyPath, ascending: sortDescriptor.ascending)
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:265]: Rather than duplicating all this logic in a whole bunch of places there should just be a single function that updates things which is called in each didSet.

File: RealmSwift/SwiftUI.swift [hunk 14]
```diff
@@ -0,0 +1,619 @@
...
+    typealias Index = Value.Index
+    /// :nodoc:
+    typealias Indices = Value.Indices
+
+    private func safeWrite(_ value: Value, _ block: () -> Void) {
+        if value.realm?.isInWriteTransaction == false {
+            value.realm?.beginWrite()
+        }
+        block()
+        try? value.realm?.commitWrite()
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:465]: try? is bad and this should only commit if it started the write

File: RealmSwift/Tests/ListTests.swift [hunk 15]
```diff
@@ -301,6 +301,42 @@ class ListTests: TestCase {
         assertThrows(array.move(from: 2, to: 0))
     }
 
+    func testMoveFromOffsets() {
```
Comment (tgoyne) [inline RealmSwift/Tests/ListTests.swift:304]: This should also test zero offsets and more than one.

File: CHANGELOG.md (hunk 12 above)
Comment (jsflax) [reply CHANGELOG.md]: fixed

File: examples/installation/osx/swift/DynamicExample/DynamicExample.xcodeproj/project.pbxproj [hunk 16]
```diff
@@ -305,7 +305,7 @@
 				DEBUG_INFORMATION_FORMAT = dwarf;
 				ENABLE_STRICT_OBJC_MSGSEND = YES;
 				ENABLE_TESTABILITY = YES;
-				FRAMEWORK_SEARCH_PATHS = "../../../realm-swift-latest/osx/swift-11.3";
+				FRAMEWORK_SEARCH_PATHS = "../../../realm-swift-latest/12.3/osx/swift-11.3";
```
Comment (tgoyne) [inline examples/installation/osx/swift/DynamicExample/DynamicExample.xcodeproj/project.pbxproj:308]: This path doesn't exist.

File: scripts/pr-ci-matrix.rb [hunk 17]
```diff
@@ -47,6 +47,7 @@ def minimum_version(major)
   'cocoapods-ios-dynamic' => oldest_and_latest,
   'cocoapods-watchos' => oldest_and_latest,
   # 'cocoapods-catalyst' => oldest_and_latest,
+  'swiftui' => latest,
```
Comment (tgoyne) [inline scripts/pr-ci-matrix.rb:50]: You also need to run this script to populate the file that jenkins actually looks at.

File: RealmSwift/EmbeddedObject.swift [hunk 18]
```diff
@@ -165,62 +165,6 @@ extension EmbeddedObject: RealmCollectionValue {
         return RLMDynamicGet(self, prop)
     }
 
-    // MARK: Notifications
```
[the remark this answers is not in the document]
Comment (jsflax) [reply RealmSwift/EmbeddedObject.swift:168]: There doesn't seem to be a clean way to do this from an extension.

[the remark this answers is not in the document]
Comment (tgoyne) [reply RealmSwift/EmbeddedObject.swift:168]: The public thing needs to be on Object/EmbeddedObject which then could call an internal function on the ObjectBase extension rather than putting the public thing on ObjectBase.

File: Realm/Tests/SwiftUITestHostUITests/SwiftUITestHostUITests.swift [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+    dynamic var name = "New List"
+    dynamic var icon: String = "list.bullet"
+    var reminders = RealmSwift.List<Reminder>()
+}
+
+class SwiftUITests: XCTestCase {
+    var realm: Realm!
+    let app = XCUIApplication()
+
+    override func setUpWithError() throws {
```
[the remark this answers is not in the document]
Comment (jsflax) [reply Realm/Tests/SwiftUITestHostUITests/SwiftUITestHostUITests.swift:28]: I excluded the file using swift version guards and made the tests themselves guarded by the `available` directive.

**Comment by jsflax** [comment]
@tgoyne , addressed all comments– still haven't come up with a better name than `FetchRealmResults` though. Added a conditional in `Realm.add` that checks if SwiftUI KVO is active on an unmanaged objects and removes if so. Added a test for this as well. PTAL.

File: RealmSwift/SwiftUI.swift [hunk 20]
```diff
@@ -121,17 +133,12 @@ private final class ObservableStoragePublisher<ObjectType>: Publisher where Obje
...
             var keyPaths = [String]()
-            for index in 0..<outCount {
-                let property = class_getProperty(ObjectType.self as? AnyClass,
-                                                 property_getName(propertyList!.advanced(by: Int(index)).pointee))
-                let name = String(cString: property_getName(property!))
-                keyPaths.append(name)
-                value.addObserver(kvo, forKeyPath: name, options: .new, context: nil)
+            for property in schema.properties {
+                keyPaths.append(property.name)
+                value.addObserver(kvo, forKeyPath: property.name, options: .new, context: nil)
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:141]: This shouldn't be asking for the new value as it doesn't use it.

File: RealmSwift/SwiftUI.swift [hunk 21]
```diff
@@ -259,56 +266,37 @@ private class ObservableStorage<ObservedType>: ObservableObject where ObservedTy
...
-                    }
-                }
+        private func didSet() {
+            /// A base value to reset the state of the query if a user reassigns the `filter` or `sortDescriptor`
+            value = try! Realm(configuration: configuration ?? Realm.Configuration.defaultConfiguration).objects(ResultType.self)
+
+            if let sortDescriptor = sortDescriptor {
+                value = value.sorted(byKeyPath: sortDescriptor.keyPath, ascending: sortDescriptor.ascending)
+            } else {
+                value = value.sorted(by: [])
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:276]: Why?

File: scripts/.jenkins.yml [hunk 22]
```diff
@@ -0,0 +1,281 @@
+
+# Yaml Axis Plugin
```
Comment (tgoyne) [inline scripts/.jenkins.yml:2]: This is the wrong path for this file.

**Review by tgoyne — COMMENTED** [review]
changelog merge is still incorrect

**Comment by jsflax** [comment]
I fixed the CHANGELOG (🤞), renamed `FetchRealmResults` to `ObservedResults`, renamed `KVO` to `SwiftUIKVO`, moved `.jenkins.yml` file to root, and added an `_objectSchema` method to `ObjectBase` as that had been missing.

**Comment by tgoyne** [comment]
There isn't really a reason to add _objectSchema; just using RLMObjectBaseObjectSchema() directly is basically the same code but with less copying involved.

**Comment by jsflax** [comment]
I'll revert that. I'm having a bit of difficulty understanding why the swiftly tests continue to fail– https://ci.realm.io/job/objc_pr/target=swiftpm,xcode_version=12.1/7669/console

Any thoughts on why the guards I've added aren't sufficient?

**Comment by tgoyne** [comment]
StateObject requires Xcode 12.2, which is Swift 5.3.1

**Comment by jsflax** [comment]
Screenshot 2021-02-12 at 18 38 56

@tgoyne our first passing UI tests 😄 . Let me know if you have any more thoughts

File: build.sh [hunk 23]
```diff
@@ -668,7 +668,9 @@ case "$COMMAND" in
         ;;
 
     "test-swiftui")
-        xctest "-scheme SwiftUITestHost -configuration $CONFIGURATION -sdk iphonesimulator -destination 'name=iPhone 12'"
+        sh build.sh download-core
```
Comment (tgoyne) [inline build.sh:671]: This should be in the scheme's pre-build action, not here.

File: Configuration/SwiftUITestHost.xctestplan [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+    {
+      "id" : "AC613D2C-06BC-4F40-AE71-3AEB7D85C05F",
+      "name" : "Configuration 1",
+      "options" : {
+
+      }
+    }
+  ],
+  "defaultOptions" : {
+    "mainThreadCheckerEnabled" : false
```
Comment (tgoyne) [inline Configuration/SwiftUITestHost.xctestplan:12]: We want the main thread checker enabled for this (and UI tests are an example of a case where it might even be useful). It's only disabled in the Xcode 11 version of the test plans, not the default one.

File: Realm/Tests/SwiftUITestHost/Objects.swift [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+    dynamic var isFlagged = false
+    dynamic var date = Date()
+    dynamic var isComplete = false
+    dynamic var priority: Priority = .low
+}
+
+@objcMembers class ReminderList: Object, ObjectKeyIdentifiable {
+    dynamic var name = "New List"
+    dynamic var icon: String = "list.bullet"
+    var reminders = RealmSwift.List<Reminder>()
```
Comment (tgoyne) [inline Realm/Tests/SwiftUITestHost/Objects.swift:47]:
```suggestion
    let reminders = RealmSwift.List<Reminder>()
```

File: Realm/Tests/SwiftUITestHost/SwiftUITestHostApp.swift [hunk 26]
```diff
@@ -0,0 +1,255 @@
...
+            DatePicker("date", selection: $reminder.date)
+            Picker("priority", selection: $reminder.priority, content: {
+                ForEach(Reminder.Priority.allCases) { priority in
+                    Text(priority.description).tag(priority)
+                }
+            }).accessibilityIdentifier("picker")
+        }
+        .navigationTitle(reminder.title)
+        .navigationBarItems(trailing:
+        Button("Save") {
```
Comment (tgoyne) [inline Realm/Tests/SwiftUITestHost/SwiftUITestHostApp.swift:54]:
```suggestion
        .navigationBarItems(trailing: Button("Save") {
```

File: Realm/Tests/SwiftUITestHost/SwiftUITestHostApp.swift [hunk 27]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+struct ReminderListView: View {
+    @ObservedRealmObject var list: ReminderList
+    @State var newReminderAdded = false
+    @State var showReminderForm = false
+
+    func shouldFocusReminder(_ reminder: Reminder) -> Bool {
+        return newReminderAdded &&
+            list.reminders.lastIndex(of: reminder) == (list.reminders.count - 1)
```
Comment (tgoyne) [inline Realm/Tests/SwiftUITestHost/SwiftUITestHostApp.swift:70]:
```suggestion
        return newReminderAdded && list.reminders.last == reminder
```

File: RealmSwift/Object.swift [hunk 28]
```diff
@@ -212,11 +212,10 @@ extension Object: RealmCollectionValue {
...
 
-     If no queue is given, notifications are delivered via the standard run
-     loop, and so can't be delivered while the run loop is blocked by other
-     activity. If a queue is given, notifications are delivered to that queue
-     instead. When notifications can't be delivered instantly, multiple
-     notifications may be coalesced into a single notification.
+     Notifications are delivered via the standard run loop, and so can't be
+     delivered while the run loop is blocked by other activity. When
+     notifications can't be delivered instantly, multiple notifications may be
+     coalesced into a single notification.
```
Comment (tgoyne) [inline RealmSwift/Object.swift:218]: The previous version of this wasn't great, but removing the mention of the queue isn't an improvement to it.

File: RealmSwift/SwiftUI.swift [hunk 29]
```diff
@@ -0,0 +1,599 @@
...
+    if didStartWrite {
+        try! value.realm?.commitWrite()
+    }
+}
+
+@available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
+private func createBinding<T: ThreadConfined, V>(_ value: T,
+                                                 forKeyPath keyPath: ReferenceWritableKeyPath<T, V>) -> Binding<V> {
+    if value.isFrozen {
+        throwRealmException("Should not bind frozen value")
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:43]: This appears to be something that's supposed to be unreachable. If that's correct, it should be a `precondition()` to make that clear, while if it's something the user actually can hit then it should be a bit more descriptive.

File: RealmSwift/SwiftUI.swift [hunk 30]
```diff
@@ -0,0 +1,599 @@
...
+    }
+    // store last known value outside of the binding so that we can reference it if the parent
+    // is invalidated
+    let lastValue = value[keyPath: keyPath]
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:47]: Shouldn't lastValue be updated each time the value changes? If that doesn't make sense to do then it should be `initialValue`.

File: RealmSwift/SwiftUI.swift [hunk 31]
```diff
@@ -0,0 +1,599 @@
...
+                sortDescriptor: SortDescriptor? = nil) {
+        self.storage.configuration = configuration
+        self.filter = filter
+        self.sortDescriptor = sortDescriptor
+    }
+
+    public mutating func update() {
+        // When the view updates, it will inject the @Environment
+        // into the propertyWrapper
+        if storage.configuration == nil ||
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:336]: Making `Realm.Configuration` conform to Equatable and putting this logic there seems better.

File: RealmSwift/SwiftUI.swift [hunk 32]
```diff
@@ -0,0 +1,599 @@
...
+    /// :nodoc:
+    typealias Element = Value.Element
+    /// :nodoc:
+    typealias Index = Value.Index
+    /// :nodoc:
+    typealias Indices = Value.Indices
+
+    /// :nodoc:
+    func remove<V>(at index: Index) where Value == List<V> {
+        let list = self.wrappedValue.realm != nil ?
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:444]: This logic shows up in a lot of places and would be worth extracting to a function.

File: RealmSwift/Tests/SwiftUITests.swift [hunk 33]
```diff
@@ -0,0 +1,239 @@
...
+        XCTAssertEqual(state.wrappedValue.count, 1)
+
+        state.projectedValue.remove(at: 0)
+        XCTAssertEqual(state.wrappedValue.count, 0)
+    }
+
+    // MARK: - ObservedResults Operations
+    func testResultsAppendUnmanagedObject() throws {
+        let object = SwiftUIObject()
+        let fullResults = ObservedResults(SwiftUIObject.self,
```
Comment (tgoyne) [inline RealmSwift/Tests/SwiftUITests.swift:133]: Indentation is off in a bunch of spots after the renames.

**Review by tgoyne — COMMENTED** [review]
Took another pass over the whole thing. Mostly looks good and all my remaining comments are minor things.

File: Realm/Tests/SwiftUITestHost/Objects.swift (hunk 25 above)
Comment (jsflax) [reply Realm/Tests/SwiftUITestHost/Objects.swift:47]: I've had to unapply this suggestion– to be able to use the `$` convenience methods, it needs to be mutable.

**Review by tgoyne — APPROVED** [review]
